File: baek_1260.cpp
#include <stdio.h>
#include <algorithm>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

vector<bool> visited;
vector<vector<int>> node;
vector<int> idx;

void dfs(int start){
    stack<int> stk;
    int cur, esize;
    visited[start] = true;
    stk.push(start);
    while(!stk.empty()){
        cur = stk.top();
        if(idx[cur]==-1){
            printf("%d ", cur);
            idx[cur]++;
        }
        esize = node[cur].size();
        for( ; idx[cur]<esize; idx[cur]++){
            if(!visited[node[cur][idx[cur]]]){
                visited[node[cur][idx[cur]]] = true;
                stk.push(node[cur][idx[cur]]);
                break;
            }
        }
        if(idx[cur]==esize)
            stk.pop();
    }
}
void bfs(int start){
    queue<int> q;
    int cur, esize;
    q.push(start);
    visited[start] = true;
    while(!q.empty()){
        cur = q.front();
        q.pop();
        printf("%d ", cur);
        esize = node[cur].size();
        for(int i=0; i<esize; i++){
            if(!visited[node[cur][i]]){
                visited[node[cur][i]] = true;
                q.push(node[cur][i]);
            }
        }
    }

}

bool cmp(int a, int b){
    return a > b;
}

int main(void){
    int n, m, v, a, b;
    scanf("%d %d %d", &n, &m, &v);
    visited.assign(n+1, false);
    idx.assign(n+1, -1);
    node.assign(n+1, vector<int>(0, 0));
    for(int i=0; i<m; i++){
        scanf("%d %d", &a, &b);
        node[a].push_back(b);
        node[b].push_back(a);
    }
    for(int i=1; i<=n; i++)
        sort(node[i].begin(), node[i].end());
    dfs(v);
    printf("\n");
    for(int i=1; i<=n; i++)
        visited[i] = false;
    bfs(v);
    printf("\n");
    return 0;
}

File: baek_1260.java
import java.util.*;

public class Main {
    static List<List<Integer>> linkedList;

    static List<Integer> dfs(int start) throws Exception {
        int vertexCount = linkedList.size() - 1;
        int cur, tar, elen;
        boolean[] visited = new boolean[vertexCount+1];
        int[] idx = new int[vertexCount+1];
        Stack<Integer> stk = new Stack<>();
        List<Integer> vList = new ArrayList<>();
        Arrays.fill(visited, false);
        Arrays.fill(idx, -1);
        stk.push(start);
        visited[start] = true;
        while(!stk.empty()){
            cur = stk.peek();
            if(idx[cur]==-1){
                vList.add(cur);
                idx[cur]++;
            }
            elen = linkedList.get(cur).size();
            for( ; idx[cur]<elen; idx[cur]++){
                tar = linkedList.get(cur).get(idx[cur]);
                if(!visited[tar]){
                    visited[tar] = true;
                    stk.push(tar);
                    break;
                }
            }
            if(idx[cur]==linkedList.get(cur).size())
                stk.pop();
        }
        return vList;
    }
    static List<Integer> bfs(int start){
        int vertexCount = linkedList.size() - 1;
        int cur, tar, elen;
        List<Integer> vList = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();
        boolean[] visited = new boolean[vertexCount+1];
        Arrays.fill(visited, false);
        queue.offer(start);
        visited[start] = true;
        vList.add(start);
        while(!queue.isEmpty()){
            cur = queue.poll();
            elen = linkedList.get(cur).size();
            for(int i=0; i<elen; i++){
                tar = linkedList.get(cur).get(i);
                if(!visited[tar]){
                    vList.add(tar);
                    visited[tar] = true;
                    queue.offer(tar);
                }
            }
        }
        return vList;
    }

    public static void main(String[] args) throws Exception{
        Scanner sc = new Scanner(System.in);
        int n, m, v, from, to, dfsLen, bfsLen;
        linkedList = new ArrayList<>();
        n = sc.nextInt();
        m = sc.nextInt();
        v = sc.nextInt();
        for(int i=0; i<=n; i++)
            linkedList.add(new ArrayList<>());
        for(int i=0; i<m; i++){
            from = sc.nextInt();
            to = sc.nextInt();
            linkedList.get(from).add(to);
            linkedList.get(to).add(from);
        }
        for(int i=1; i<=n; i++)
            Collections.sort(linkedList.get(i));
        List<Integer> dfsList = dfs(v);
        List<Integer> bfsList = bfs(v);
        dfsLen = dfsList.size();
        bfsLen = bfsList.size();
        for(int i=0; i<dfsLen; i++){
            System.out.print(dfsList.get(i));
            System.out.print(" ");
        }
        System.out.println();
        for(int i=0; i<bfsLen; i++){
            System.out.print(bfsList.get(i));
            System.out.print(" ");
        }
    }
}
